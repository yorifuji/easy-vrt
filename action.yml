name: "Easy VRT"

description: "Easy VRT is a utility action for visual regression testing."

author: "yorifuji"

branding:
  icon: code
  color: blue

inputs:
  mode:
    description: "Mode, [lookup, expected, actual, compare]"
    required: true
  expected-dir:
    description: "Expected directory"
    required: false
  expected-cache-key:
    description: "Cache key for expected"
    required: false
  actual-dir:
    description: "Actual directory"
    required: false
  actual-cache-key:
    description: "Cache key for actual"
    required: false
  summary-comment:
    description: "If true, add a summary comment on workflow summary"
    required: false
    default: false
  review-comment:
    description: "If true, add a review comment on PR review"
    required: false
    default: false

outputs:
  actual-sha:
    description: "Actual SHA"
    value: ${{ steps.lookup-sha.outputs.actual-sha }}
  actual-cache-hit:
    description: "Whether the actual cache was found"
    value: ${{ steps.lookup-actual-cache.outputs.cache-hit }}
  expected-sha:
    description: "Expected SHA"
    value: ${{ steps.lookup-sha.outputs.expected-sha }}
  expected-cache-hit:
    description: "Whether the expected cache was found"
    value: ${{ steps.lookup-expected-cache.outputs.cache-hit }}

runs:
  using: composite
  steps:
    - shell: bash
      run: |
        echo "mode: ${{ inputs.mode }}"

    # lookup

    - if: ${{ inputs.mode == 'lookup' }}
      uses: actions/checkout@v4
      with:
        fetch-depth: 0

    - if: ${{ inputs.mode == 'lookup' }}
      id: lookup-sha
      shell: bash
      run: |
        echo "actual-sha=${{ github.event.pull_request.head.sha }}" >> $GITHUB_OUTPUT
        echo "expected-sha=$(git merge-base ${{ github.event.pull_request.base.sha }} ${{ github.event.pull_request.head.sha }})" >> $GITHUB_OUTPUT

    - if: ${{ inputs.mode == 'lookup' }}
      id: lookup-actual-cache
      uses: actions/cache@v4
      with:
        key: reg-suit-cache-${{ steps.lookup-sha.outputs.actual-sha }}
        path: .easy-vrt/actual
        lookup-only: true

    - if: ${{ inputs.mode == 'lookup' }}
      id: lookup-expected-cache
      uses: actions/cache@v4
      with:
        key: reg-suit-cache-${{ steps.lookup-sha.outputs.expected-sha }}
        path: .easy-vrt/expected
        lookup-only: true

    # expected

    - if: ${{ inputs.mode == 'expected' }}
      shell: bash
      run: |
        if [ -e .easy-vrt ]; then exit 1; fi
        mkdir .easy-vrt && mv ${{ inputs.expected-dir }} .easy-vrt/expected

    - if: ${{ inputs.mode == 'expected' }}
      uses: actions/cache/save@v4
      with:
        path: .easy-vrt/expected
        key: reg-suit-cache-${{ inputs.expected-cache-key }}

    # actual

    - if: ${{ inputs.mode == 'actual' }}
      shell: bash
      run: |
        if [ -e .easy-vrt ]; then exit 1; fi
        mkdir .easy-vrt && mv ${{ inputs.actual-dir }} .easy-vrt/actual

    - if: ${{ inputs.mode == 'actual' }}
      uses: actions/cache/save@v4
      with:
        path: .easy-vrt/actual
        key: reg-suit-cache-${{ inputs.actual-cache-key }}

    # compare

    - if: ${{ inputs.mode == 'compare' }}
      shell: bash
      run: |
        echo expected-cache-key ${{ inputs.expected-cache-key }}
        echo actual-cache-key ${{ inputs.actual-cache-key }}

    - if: ${{ inputs.mode == 'compare' }}
      shell: bash
      run: |
        if [ -e .easy-vrt ]; then exit 1; fi

    - if: ${{ inputs.mode == 'compare' }}
      shell: bash
      run: |
        if [ -e package.json ]; then exit 1; fi
        if [ -e package-lock.json ]; then exit 1; fi
        cp $GITHUB_ACTION_PATH/package.json $GITHUB_ACTION_PATH/package-lock.json .

    - if: ${{ inputs.mode == 'compare' }}
      uses: actions/setup-node@v4
      with:
        node-version: 20
        cache: npm

    - if: ${{ inputs.mode == 'compare' }}
      shell: bash
      run: npm install

    - if: ${{ inputs.mode == 'compare' }}
      uses: actions/cache/restore@v4
      with:
        key: reg-suit-cache-${{ inputs.actual-cache-key }}
        path: .easy-vrt/actual

    - if: ${{ inputs.mode == 'compare' }}
      uses: actions/cache/restore@v4
      with:
        key: reg-suit-cache-${{ inputs.expected-cache-key }}
        path: .easy-vrt/expected

    - if: ${{ inputs.mode == 'compare' }}
      shell: bash
      run: >
        npx reg-cli
        .easy-vrt/actual
        .easy-vrt/expected
        .easy-vrt/diff
        -R .easy-vrt/report.html
        -J .easy-vrt/report.json

    - if: ${{ !cancelled() && inputs.mode == 'compare' }}
      id: upload-artifact
      uses: actions/upload-artifact@v4
      with:
        name: easy-vrt-report
        path: .easy-vrt
        # retention-days: 3

    - if: ${{ !cancelled() && inputs.mode == 'compare' }}
      uses: actions/github-script@v7
      env:
        SUMMARY_COMMENT: ${{ inputs.summary-comment }}
        REVIEW_COMMENT: ${{ inputs.review-comment }}
        ARTIFACT_URL: ${{ steps.upload-artifact.outputs.artifact-url }}
      with:
        script: |
          const fs = require('fs');

          const summaryComment = process.env.SUMMARY_COMMENT === 'true';
          const reviewComment = process.env.REVIEW_COMMENT === 'true';
          const artifactUrl = process.env.ARTIFACT_URL;

          const log = fs.readFileSync('.easy-vrt/report.json', 'utf-8');
          console.log(log);
          const json = JSON.parse(log);
          console.log(json);

          const titleIcon = 'âœ…';
          const easyVrtComment = '<!-- Easy VRT Comment -->';

          const stats = {
            changed: json.failedItems.length.toString(),
            newItems: json.newItems.length.toString(),
            deleted: json.deletedItems.length.toString(),
            passing: json.passedItems.length.toString()
          };

          const markdown = await core.summary
            .addHeading(`${titleIcon} Easy VRT has checked for visual changes`, 3)
            .addTable([
              ["ðŸ”´ Changed",  "ðŸŸ¡ New",        "ðŸŸ¤  Deleted",  "ðŸ”µ Passing"],
              [stats.changed, stats.newItems, stats.deleted, stats.passing]
            ])
            .addHeading("ðŸ“Š Download Report", 3)
            .addLink('You can download the report from the artifact here', `${artifactUrl}`)
            .addRaw(`${easyVrtComment}`)
            .stringify()

          if (summaryComment) {
            await core.summary.write()
          }

          if (reviewComment) {
            const requestPerPage = 100;
            try {
              const listComments = await github.rest.issues.listComments({
                owner: context.repo.owner,
                repo: context.repo.repo,
                issue_number: context.issue.number,
                per_page: requestPerPage
              });
              const easyVrtCommentList = listComments.data.find(comment => comment.body.includes(easyVrtComment));
              if (easyVrtCommentList) {
                // delete summary comment if exists
                await github.rest.issues.deleteComment({
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  comment_id: easyVrtCommentList.id
                });
              }
              // create a comment
              await github.rest.issues.createComment({
                owner: context.repo.owner,
                repo: context.repo.repo,
                issue_number: context.issue.number,
                body: markdown
              });
            } catch (error) {
              core.error(`Failed to modify comment: ${error.message}`);
              return;
            }
          }
