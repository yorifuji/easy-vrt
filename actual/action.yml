name: "Actual VRT"

description: "Actual VRT is a utility action for visual regression testing."

author: "yorifuji"

branding:
  icon: code
  color: blue

inputs:
  actual-dir:
    description: "Actual directory"
    required: true
  actual-cache-key:
    description: "Cache key for actual"
    required: true

runs:
  using: composite
  steps:
    - shell: bash
      run: |
        if [ -e .easy-vrt ]; then exit 1; fi
        mkdir .easy-vrt && mv ${{ inputs.actual-dir }} .easy-vrt/actual

    - uses: actions/cache/save@d4323d4df104b026a6aa633fdb11d772146be0bf # v4.1.2
      with:
        path: .easy-vrt/actual
        key: reg-suit-cache-${{ inputs.actual-cache-key }}
