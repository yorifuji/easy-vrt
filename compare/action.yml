name: "Compare VRT"

description: "Compare VRT is a utility action for visual regression testing."

author: "yorifuji"

branding:
  icon: code
  color: blue

inputs:
  expected-cache-key:
    description: "Cache key for expected"
    required: required
  actual-cache-key:
    description: "Cache key for actual"
    required: required
  summary-comment:
    description: "If true, add a summary comment on workflow summary"
    required: false
    default: false
  review-comment:
    description: "If true, add a review comment on PR review"
    required: false
    default: false

runs:
  using: composite
  steps:
    - shell: bash
      run: |
        echo expected-cache-key ${{ inputs.expected-cache-key }}
        echo actual-cache-key ${{ inputs.actual-cache-key }}

    - shell: bash
      run: |
        if [ -e .easy-vrt ]; then exit 1; fi

    - shell: bash
      run: |
        if [ -e package.json ]; then exit 1; fi
        if [ -e package-lock.json ]; then exit 1; fi
        cp $GITHUB_ACTION_PATH/package.json $GITHUB_ACTION_PATH/package-lock.json .

    - uses: actions/setup-node@39370e3970a6d050c480ffad4ff0ed4d3fdee5af # v4.1.0
      with:
        node-version: 20
        cache: npm

    - shell: bash
      run: npm install

    - uses: actions/cache/restore@6849a6489940f00c2f30c0fb92c6274307ccb58a # v4.1.2
      with:
        key: reg-suit-cache-${{ inputs.actual-cache-key }}
        path: .easy-vrt/actual

    - uses: actions/cache/restore@6849a6489940f00c2f30c0fb92c6274307ccb58a # v4.1.2
      with:
        key: reg-suit-cache-${{ inputs.expected-cache-key }}
        path: .easy-vrt/expected

    - shell: bash
      run: >
        npx reg-cli
        .easy-vrt/actual
        .easy-vrt/expected
        .easy-vrt/diff
        -R .easy-vrt/report.html
        -J .easy-vrt/report.json
        --ignoreChange

    - if: ${{ !cancelled() }}
      id: upload-artifact
      uses: actions/upload-artifact@b4b15b8c7c6ac21ea08fcf65892d2ee8f75cf882 # v4.4.3
      with:
        name: easy-vrt-report
        path: .easy-vrt
        include-hidden-files: true
        # retention-days: 3

    - if: ${{ !cancelled() }}
      uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7.0.1
      env:
        SUMMARY_COMMENT: ${{ inputs.summary-comment }}
        REVIEW_COMMENT: ${{ inputs.review-comment }}
        ARTIFACT_URL: ${{ steps.upload-artifact.outputs.artifact-url }}
      with:
        script: |
          const fs = require('fs');

          const summaryComment = process.env.SUMMARY_COMMENT === 'true';
          const reviewComment = process.env.REVIEW_COMMENT === 'true';
          const artifactUrl = process.env.ARTIFACT_URL;

          const log = fs.readFileSync('.easy-vrt/report.json', 'utf-8');
          console.log(log);
          const json = JSON.parse(log);
          console.log(json);

          const titleIcon = 'âœ…';
          const easyVrtComment = '<!-- Easy VRT Comment -->';

          const stats = {
            changed: json.failedItems.length.toString(),
            newItems: json.newItems.length.toString(),
            deleted: json.deletedItems.length.toString(),
            passing: json.passedItems.length.toString()
          };

          const markdown = await core.summary
            .addHeading(`${titleIcon} Easy VRT has checked for visual changes`, 3)
            .addTable([
              ["ðŸ”´ Changed",  "ðŸŸ¡ New",        "ðŸŸ¤  Deleted",  "ðŸ”µ Passing"],
              [stats.changed, stats.newItems, stats.deleted, stats.passing]
            ])
            .addHeading("ðŸ“Š Download Report", 3)
            .addLink('You can download the entire report here', `${artifactUrl}`)
            .addRaw(`${easyVrtComment}`)
            .stringify()

          if (summaryComment) {
            await core.summary.write()
          }

          if (reviewComment) {
            const requestPerPage = 100;
            try {
              const listComments = await github.rest.issues.listComments({
                owner: context.repo.owner,
                repo: context.repo.repo,
                issue_number: context.issue.number,
                per_page: requestPerPage
              });
              const easyVrtCommentList = listComments.data.find(comment => comment.body.includes(easyVrtComment));
              if (easyVrtCommentList) {
                // delete summary comment if exists
                await github.rest.issues.deleteComment({
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  comment_id: easyVrtCommentList.id
                });
              }
              // create a comment
              await github.rest.issues.createComment({
                owner: context.repo.owner,
                repo: context.repo.repo,
                issue_number: context.issue.number,
                body: markdown
              });
            } catch (error) {
              core.error(`Failed to modify comment: ${error.message}`);
              return;
            }
          }
