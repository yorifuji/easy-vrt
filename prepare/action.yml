name: "Prepare VRT"

description: "Prepare VRT is a utility action for visual regression testing."

author: "yorifuji"

branding:
  icon: code
  color: blue

outputs:
  actual-sha:
    description: "Actual SHA"
    value: ${{ steps.lookup-sha.outputs.actual-sha }}
  actual-cache-hit:
    description: "Whether the actual cache was found"
    value: ${{ steps.lookup-actual-cache.outputs.cache-hit }}
  expected-sha:
    description: "Expected SHA"
    value: ${{ steps.lookup-sha.outputs.expected-sha }}
  expected-cache-hit:
    description: "Whether the expected cache was found"
    value: ${{ steps.lookup-expected-cache.outputs.cache-hit }}

runs:
  using: composite
  steps:
    - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
      with:
        fetch-depth: 0

    - id: lookup-sha
      shell: bash
      run: |
        echo "actual-sha=${{ github.event.pull_request.head.sha }}" >> $GITHUB_OUTPUT
        echo "expected-sha=$(git merge-base ${{ github.event.pull_request.base.sha }} ${{ github.event.pull_request.head.sha }})" >> $GITHUB_OUTPUT

    - id: lookup-actual-cache
      uses: actions/cache@5a3ec84eff668545956fd18022155c47e93e2684 # v4.1.2
      with:
        key: reg-suit-cache-${{ steps.lookup-sha.outputs.actual-sha }}
        path: .easy-vrt/actual
        lookup-only: true

    - id: lookup-expected-cache
      uses: actions/cache@5a3ec84eff668545956fd18022155c47e93e2684 # v4.1.2
      with:
        key: reg-suit-cache-${{ steps.lookup-sha.outputs.expected-sha }}
        path: .easy-vrt/expected
        lookup-only: true
