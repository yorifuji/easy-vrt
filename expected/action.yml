name: "Expected VRT"

description: "Expected VRT is a utility action for visual regression testing."

author: "yorifuji"

branding:
  icon: code
  color: blue

inputs:
  expected-dir:
    description: "Expected directory"
    required: true
  expected-cache-key:
    description: "Cache key for expected"
    required: true

runs:
  using: composite
  steps:
    # expected

    - shell: bash
      run: |
        if [ -e .easy-vrt ]; then exit 1; fi
        mkdir .easy-vrt && mv ${{ inputs.expected-dir }} .easy-vrt/expected

    - uses: actions/cache/save@5a3ec84eff668545956fd18022155c47e93e2684 # v4.1.2
      with:
        path: .easy-vrt/expected
        key: reg-suit-cache-${{ inputs.expected-cache-key }}
